src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -301,32 +260,47 @@ private AddTxResult AddCore(Transaction tx, bool isPersistentBroadcast, bool isR
                     _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, nonce already used.");
                 return AddTxResult.OldNonce;
             }
-
-            if (tx.Nonce > currentNonce + FutureNonceRetention)
+            
+            int numberOfSenderTxsInPending = _transactions.GetBucketCount(tx.SenderAddress);
+            bool isTxPoolFull = GetPendingTransactionsCount() == MemoryAllowance.MemPoolSize;
+            bool isTxNonceNextInOrder = tx.Nonce <= (long)currentNonce + numberOfSenderTxsInPending;
+            bool isTxNonceTooFarInFuture = tx.Nonce > currentNonce + FutureNonceRetention;
+            if (isTxPoolFull && !isTxNonceNextInOrder
+                || isTxNonceTooFarInFuture)
             {
                 if (_logger.IsTrace)
-                    _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, nonce in far future.");
+                    _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, nonce in future.");
                 return AddTxResult.FutureNonce;
             }
 
-            
+            UInt256 balance = account?.Balance ?? UInt256.Zero;
+            UInt256 payableGasPrice = tx.CalculatePayableGasPrice(spec.IsEip1559Enabled, CurrentBaseFee, balance);
+
             bool overflow = spec.IsEip1559Enabled && UInt256.AddOverflow(tx.MaxPriorityFeePerGas, tx.MaxFeePerGas, out _);
             // we're checking that user can pay what he declared in FeeCap. For this check BaseFee = FeeCap
-            UInt256 effectiveGasPrice = tx.CalculateEffectiveGasPrice(spec.IsEip1559Enabled, tx.MaxFeePerGas);
-            overflow |= UInt256.MultiplyOverflow(effectiveGasPrice, (UInt256) tx.GasLimit, out UInt256 cost);
+            overflow |= UInt256.MultiplyOverflow(payableGasPrice, (UInt256) tx.GasLimit, out UInt256 cost);
             overflow |= UInt256.AddOverflow(cost, tx.Value, out cost);
             if (overflow)
             {
                 if (_logger.IsTrace)
                     _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, cost overflow.");
                 return AddTxResult.BalanceOverflow;
             }
-            else if ((account?.Balance ?? UInt256.Zero) < cost)
+            else if (balance < cost)
             {
                 if (_logger.IsTrace)
                     _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, insufficient funds.");
                 return AddTxResult.InsufficientFunds;
             }
+                
+            if (isTxPoolFull
+                && _transactions.TryGetLast(out var lastTx)
+                && payableGasPrice <= lastTx?.GasBottleneck)
+            {
+                if (_logger.IsTrace)
+                    _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, too low gasPrice.");
```

Comment:
"Too low gas premium" is inaccurate, because for 1559 payableGasPrice can be low while gas premium is high (because of low FeeCap). Even "too low effective gas price" is inaccurate, because payableGasPrice can be limited by account balance. I think we can use "too low payable gas price", or just "too low fee".

Reply:
ok discuss with @MarekM25 best naming